### ISSUE TITLE
TakeBackward taking a significant portion of backward time

### ISSUE THREAD

**Issue by rachtsingh** [pytorch/pytorch#5159]
I'm trying to write some code that requires reindexing the hidden states of an LSTM at each step (because of particle filtering). Is there a more efficient way to do this? It's pretty slow right now, and I wonder if another approach (`scatter_`?) might be better.

If there's an internal issue with PyTorch, I'm happy to tackle it and make a PR; I just don't know where to look.

- OS: Ubuntu 16.04
- PyTorch version: `0.4.0a0+e3e3874`
- How you installed PyTorch (conda, pip, source): source
- Python version: 2.7.12
- CUDA/cuDNN version: 8.0
- GPU models and configuration: Tesla K80
- GCC version (if compiling from source): 5.4.0

```
import torch
import torch.nn as nn
from torch.autograd import Variable

INPUT_SIZE = 1024
HIDDEN_SIZE = 512


class IndexingIssues(nn.Module):
    def __init__(self):
        super(IndexingIssues, self).__init__()
        self.enc = nn.LSTMCell(INPUT_SIZE, HIDDEN_SIZE)

    def forward(self, input):
        seq_len, batch_sz, _ = input.size()
        h = Variable(input.data.new(batch_sz, HIDDEN_SIZE).zero_())
        c = Variable(input.data.new(batch_sz, HIDDEN_SIZE).zero_())
        for i in range(seq_len):
            h, c = self.enc(input[i], (h, c))
            idx = torch.arange(batch_sz).long().cuda().view(-1)  # just identity for now
            h = h[idx]
            c = c[idx]
        return h.sum()


def main():
    data = Variable(torch.randn(15, 80, INPUT_SIZE).cuda())
    model = IndexingIssues().cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=0.0001)
    with torch.autograd.profiler.profile() as prof:
        for i in range(5):
            optimizer.zero_grad()
            loss = model(data)
            loss.backward()
            optimizer.step()
    prof.export_chrome_trace("repro.prof")


if __name__ == '__main__':
    main()
```

Here's the chrome_trace it outputs [repro.txt](https://github.com/pytorch/pytorch/files/1711786/repro.txt)
(this isn't a text file, just needed to rename so Github would accept it)

**Comment by fmassa** [comment]
Did you try using `index_select`? It doesn't support broadcasting nor advanced indexing, but might be a bit faster than using the indexing operator

**Comment by rachtsingh** [comment]
Yes, you were completely right - it's 5-10x faster on backward compared to direct indexing on some benchmarks I just ran. This is really helpful, thank you very much.

### indexing operator

Name | Wall Duration | Self time | Average Duration | Occurrences
-- | -- | -- | -- | --
SumBackward0 | 0.014 ms | 0.004 ms | 0.014 ms | 1
TakeBackward | 7.619 ms | 0.378 ms | 0.263 ms | 29
LSTMFusedBackward | 3.276 ms | 2.191 ms | 0.218 ms | 15
MmBackward | 3.155 ms | 0.710 ms | 0.105 ms | 30
TBackward | 0.265 ms | 0.152 ms | 0.009 ms | 30
add | 1.705 ms | 1.705 ms | 0.030 ms | 56
N5torch8autograd14AccumulateGradE | 0.107 ms | 0.107 ms | 0.027 ms | 4
expand | 0.010 ms | 0.010 ms | 0.010 ms | 1
zeros | 0.488 ms | 0.488 ms | 0.016 ms | 30
put_ | 6.767 ms | 6.767 ms | 0.233 ms | 29
tensor | 0.202 ms | 0.202 ms | 0.007 ms | 30
sum | 0.869 ms | 0.869 ms | 0.029 ms | 30
t | 0.557 ms | 0.557 ms | 0.005 ms | 104
mm | 2.002 ms | 2.002 ms | 0.045 ms | 44
Totals | 27.036 ms | 16.142 ms | 0.062 ms | 433



### `index_select`

Name | Wall Duration | Self time | Average Duration | Occurrences
-- | -- | -- | -- | --
N5torch8autograd9GraphRootE | 0.002 ms | 0.002 ms | 0.002 ms | 1
SumBackward0 | 0.013 ms | 0.004 ms | 0.013 ms | 1
IndexSelectBackward | 1.164 ms | 0.226 ms | 0.040 ms | 29
LSTMFusedBackward | 2.568 ms | 1.817 ms | 0.171 ms | 15
MmBackward | 2.140 ms | 0.325 ms | 0.071 ms | 30
TBackward | 0.330 ms | 0.215 ms | 0.011 ms | 30
add | 1.095 ms | 1.095 ms | 0.020 ms | 56
N5torch8autograd14AccumulateGradE | 0.042 ms | 0.042 ms | 0.010 ms | 4
expand | 0.009 ms | 0.009 ms | 0.009 ms | 1
zeros | 0.509 ms | 0.509 ms | 0.017 ms | 30
index_add_ | 0.443 ms | 0.443 ms | 0.015 ms | 29
tensor | 0.209 ms | 0.209 ms | 0.007 ms | 30
sum | 0.528 ms | 0.528 ms | 0.018 ms | 30
t | 0.564 ms | 0.564 ms | 0.005 ms | 104
mm | 1.366 ms | 1.366 ms | 0.031 ms | 44
Totals | 10.982 ms | 7.353 ms | 0.025 ms | 434

I'm not sure how accurate these profiling dumps are (since I'm not sure whether they measure kernel execution time correctly), but at least in terms of wall clock time it's definitely worthwhile.

**Comment by justinchiu** [comment]
It seems weird that put_ with accumulate=True is that much slower than index_add_ in this case. Is that expected?

**Comment by apaszke** [comment]
It's a perf regression, so I'll reopen it. It's not expected. Our indexing should use the fastest ops it can

**Comment by colesbury** [comment]
A few notes:

1. Indexing a single dimension of multi-dimensional arrays is slower than it should be because it expands the indexing tensor to individual elements. This should be fixed.
2. `index_select` has an incorrect gradient when indices repeat. That's why `put` with `accumulate=True` is slower than `put` with `accumulate=False`. Perhaps we can figure out a way to specify that indices don't repeat but I think fixing (1) is more urgent.

**Comment by mbp28** [comment]
@colesbury 
Is it safe to use index_select with repeated indices in 0.4 now or has this not been fixed yet? I assume the use in 0.3 (which I am still using) is discouraged, because it definitely works incorrectly there with repeated indices?
